File: app/src/main/java/ua/syt0r/furiganit/service/OverlayTouchListener.java
package ua.syt0r.furiganit.service;

import android.content.Context;
import android.os.CountDownTimer;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import androidx.core.view.GestureDetectorCompat;

class OverlayTouchListener implements View.OnTouchListener {

    private WindowManager windowManager;

    private GestureDetectorCompat gestureDetectorCompat;
    private CountDownTimer countDownTimer;

    private int initialX;
    private int initialY;
    private float initialTouchX;
    private float initialTouchY;

    public OverlayTouchListener(Context context, WindowManager windowManager) {
        this.windowManager = windowManager;
        gestureDetectorCompat = new GestureDetectorCompat(context, new SingleTapConfirm());
    }

    public void setCountDownTimer(CountDownTimer countDownTimer) {
        this.countDownTimer = countDownTimer;
    }

    @Override
    public boolean onTouch(View view, MotionEvent motionEvent) {

        //If clicked
        if (gestureDetectorCompat.onTouchEvent(motionEvent)) {
            countDownTimer.cancel();
            countDownTimer.onFinish();
            view.performClick();
            return true;
        }

        WindowManager.LayoutParams params = (WindowManager.LayoutParams) view.getLayoutParams();

        //if dragged
        switch (motionEvent.getAction()) {
            case MotionEvent.ACTION_DOWN:
                initialX = params.x;
                initialY = params.y;
                initialTouchX = motionEvent.getRawX();
                initialTouchY = motionEvent.getRawY();
                resetTimer();
                return true;
            case MotionEvent.ACTION_UP:
                resetTimer();
                return true;
            case MotionEvent.ACTION_MOVE:
                params.x = initialX + (int) (motionEvent.getRawX() - initialTouchX);
                params.y = initialY + (int) (motionEvent.getRawY() - initialTouchY);
                windowManager.updateViewLayout(view, params);
                resetTimer();
                return true;
        }

        return false;
    }

    private void resetTimer(){
        countDownTimer.cancel();
        countDownTimer.start();
    }

    static class SingleTapConfirm extends GestureDetector.SimpleOnGestureListener {
        @Override
        public boolean onSingleTapUp(MotionEvent event) {
            return true;
        }
    }

}

File: app/src/main/java/ua/syt0r/furiganit/core/overlay/OverlayTouchListener.kt
package ua.syt0r.furiganit.core.overlay

import android.content.Context
import android.os.CountDownTimer
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import androidx.core.view.GestureDetectorCompat

class OverlayTouchListener(
        context: Context,
        private val windowManager: WindowManager
) : View.OnTouchListener {

    private val gestureDetectorCompat = GestureDetectorCompat(context, SingleTapConfirm())
    private var countDownTimer: CountDownTimer? = null

    private var initialX: Int = 0
    private var initialY: Int = 0
    private var initialTouchX: Float = 0f
    private var initialTouchY: Float = 0f

    fun setCountDownTimer(countDownTimer: CountDownTimer) {
        this.countDownTimer = countDownTimer
    }

    override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {

        //If clicked
        if (gestureDetectorCompat.onTouchEvent(motionEvent)) {
            countDownTimer?.cancel()
            countDownTimer?.onFinish()
            view.performClick()
            return true
        }

        val params = view.layoutParams as WindowManager.LayoutParams

        //if dragged
        when (motionEvent.action) {
            MotionEvent.ACTION_DOWN -> {
                initialX = params.x
                initialY = params.y
                initialTouchX = motionEvent.rawX
                initialTouchY = motionEvent.rawY
                resetTimer()
                return true
            }
            MotionEvent.ACTION_UP -> {
                resetTimer()
                return true
            }
            MotionEvent.ACTION_MOVE -> {
                params.x = initialX + (motionEvent.rawX - initialTouchX).toInt()
                params.y = initialY + (motionEvent.rawY - initialTouchY).toInt()
                windowManager.updateViewLayout(view, params)
                resetTimer()
                return true
            }
        }

        return false
    }

    private fun resetTimer() {
        countDownTimer?.cancel()
        countDownTimer?.start()
    }

    class SingleTapConfirm : GestureDetector.SimpleOnGestureListener() {
        override fun onSingleTapUp(event: MotionEvent): Boolean {
            return true
        }
    }

}